lib/reducers/recentBlocks.js

```diff
@@ -0,0 +1,57 @@
+import { ADD_RECENT_BLOCK, SET_RECENT_BLOCKS } from '../constants';
+
+const initialState = {
+  blocks: [],
+  numBlocks: 10
+};
+
+function recentBlocksReducer(state = initialState, action) {
+  const { type, payload } = action;
+  let newState = { ...state };
+
+  switch (type) {
+    case SET_RECENT_BLOCKS: {
+      if (payload.length) {
+        newState.blocks = payload;
+      }
+      return newState;
+    }
+
+    case ADD_RECENT_BLOCK: {
+      // get a copy of the blocks array
+      const blocks = newState.blocks.slice();
+      const block = payload;
+      const { numBlocks } = newState;
+
+      // skip if we already have the block for that height
+      // or recentBlocks haven't been hydrated yet
+      // reason is new block can be received multiple times
+      if (
+        (blocks && blocks.length && !blocks[block.height]) ||
+        blocks[block.height].hash !== block.hash
+      ) {
+        blocks.unshift(payload);
```

Comment:
This method is `O(n)`, could we do something like `blocks = [payload, ...blocks]`?

Reply:
Ok, i still like the more functional approach of `updated = [payload, ...blocks]` more but am not particularly opinionated